// @amplitude/experiment-node-server/dist/src/types/config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EU_SERVER_URLS = exports.CohortSyncConfigDefaults = exports.AssignmentConfigDefaults = exports.LocalEvaluationDefaults = exports.Defaults = exports.RemoteEvaluationDefaults = void 0;
/**
 Defaults for Experiment Config options

 | **Option**       | **Default**                       |
 |----------------|---------------------------------|
 | **debug**        | false                           |
 | **serverUrl**    | `"https://api.lab.amplitude.com"` |
 | **fetchTimeoutMillis**    | `10000` |
 | **fetchRetries**    | `8` |
 | **fetchRetryBackoffMinMillis**    | `500` |
 | **fetchRetryBackoffMaxMillis**    | `10000` |
 | **fetchRetryBackoffScalar**    | `1.5` |
 | **fetchRetryTimeoutMillis**    | `10000` |
 | **httpAgent** | null |

 *
 * @category Configuration
 */
exports.RemoteEvaluationDefaults = {
    debug: false,
    serverZone: 'us',
    serverUrl: 'https://api.lab.amplitude.com',
    fetchTimeoutMillis: 10000,
    fetchRetries: 8,
    fetchRetryBackoffMinMillis: 500,
    fetchRetryBackoffMaxMillis: 10000,
    fetchRetryBackoffScalar: 1.5,
    fetchRetryTimeoutMillis: 10000,
    httpAgent: null,
};
/**
 * @deprecated use {@link RemoteEvaluationDefaults}
 */
exports.Defaults = {
    debug: false,
    serverZone: 'us',
    serverUrl: 'https://api.lab.amplitude.com',
    fetchTimeoutMillis: 10000,
    fetchRetries: 8,
    fetchRetryBackoffMinMillis: 500,
    fetchRetryBackoffMaxMillis: 10000,
    fetchRetryBackoffScalar: 1.5,
    fetchRetryTimeoutMillis: 10000,
    httpAgent: null,
};
/**
 Defaults for {@link LocalEvaluationConfig} options.

 | **Option**       | **Default**                       |
 |----------------|---------------------------------|
 | **debug**        | false                           |
 | **serverUrl**    | `"https://api.lab.amplitude.com"` |
 | **flagConfigPollingIntervalMillis**    | `30000` |
 | **httpAgent** | null |

 * @category Configuration
 */
exports.LocalEvaluationDefaults = {
    debug: false,
    serverZone: 'us',
    serverUrl: 'https://api.lab.amplitude.com',
    bootstrap: {},
    flagConfigPollingIntervalMillis: 30000,
    httpAgent: null,
    streamUpdates: false,
    streamServerUrl: 'https://stream.lab.amplitude.com',
    streamFlagConnTimeoutMillis: 1500,
};
exports.AssignmentConfigDefaults = {
    cacheCapacity: 65536,
};
exports.CohortSyncConfigDefaults = {
    cohortServerUrl: 'https://cohort-v2.lab.amplitude.com',
    maxCohortSize: 2147483647,
    cohortPollingIntervalMillis: 60000,
};
exports.EU_SERVER_URLS = {
    name: 'eu',
    remote: 'https://api.lab.eu.amplitude.com',
    flags: 'https://flag.lab.eu.amplitude.com',
    stream: 'https://stream.lab.eu.amplitude.com',
    cohort: 'https://cohort-v2.lab.eu.amplitude.com',
};
